### PR TITLE
Async flag for Queue, for testing purposes

### PR DESCRIPTION
**Pull request by nuschk**

### CODE REVIEW

**Comment by jonparrott** [comment]
@nuschk thanks for your PR - we'll need you to sign the CL before I can review and merge. Let me know if you have any questions.

**Comment by jonparrott** [comment]
@nuschk I like the idea, but I'm wary on the implementation. How would you feel about just adding a new Queue class, e.g. `ImmediateQueue` or `TestQueue` or `SyncQueue`?

**Comment by nuschk** [comment]
We could certainly do that. I agree that it makes the implementation much nicer. The client code, however, gets a slight bit more complex and harder to read. Consider this:

```
is_async = settings.PSQ_ASYNC_QUEUES

if is_async:
    pubsub_client = google.cloud.pubsub.Client()
else:
    pubsub_client = None

fast = psq.Queue(pubsub=pubsub_client, name="fast", async=is_async)
slow = psq.Queue(pubsub=pubsub_client, name="slow", async=is_async)
```

vs.

```
is_async = settings.PSQ_ASYNC_QUEUES

if is_async:
    args = dict(pubsub_client=google.cloud.pubsub.Client())
    QueueClass = psq.Queue
else:
    args = dict()
    QueueClass = psq.ImmediateQueue

fast = QueueClass(name="fast", **args)
slow = QueueClass(name="slow", **args)
```

(Given that you'd want to not copy the actual queues' definitions)

rq took the first approach, which is also what nudged me into this direction. But honestly, I can live with both solutions, and don't have a strong opinion. I think it's more a question of the general goal of psq, e.g., should a migration from rq be as easy as possible. I think it's up to you, really. :-)

**Comment by jonparrott** [comment]
@nuschk that's a fair point, give me some time to review how RQ does this and I'll report back

(thanks for your patience)

**Comment by jonparrott** [comment]
(This is still on my radar, I just haven't had a chance to review rq's implementation yet. Thanks for your patience)

**Comment by nuschk** [comment]
Hehe, no worries, take your time. And thanks for the heads up.

File: psq/task.py [hunk 1]
```diff
@@ -90,6 +94,8 @@ def execute(self, queue=None):
                 self.retry()
                 queue.enqueue_task(self)
             except Exception as e:
+                logger.exception(e)
```
Comment (jonparrott) [inline psq/task.py]: just `logger.exception('Task {} failed.'.format(self.id))`, `logger.exception` automatically logs a stacktrace. :)

File: psq/queue.py [hunk 2]
```diff
@@ -34,11 +34,15 @@
 
 class Queue(object):
     def __init__(self, pubsub, name='default', storage=None,
-                 extra_context=None):
-        _check_for_thread_safety(pubsub)
-        self.pubsub = pubsub
+                 extra_context=None, async=True):
+        self.async = async
```
Comment (nuschk) [inline psq/queue.py]: Done.

File: psq/queue.py [hunk 3]
```diff
@@ -91,7 +95,12 @@ def enqueue_task(self, task):
         Note that this does not store the task.
         """
         data = dumps(task)
-        self.topic.publish(data)
+
+        if self.async:
+            self.topic.publish(data)
+        else:
+            task.execute(queue=self)
```
Comment (nuschk) [inline psq/queue.py]: Oops, yes. I've added measure_time() as well and made the logging distinctive depending on async.

File: psq/queue_test.py [hunk 4]
```diff
@@ -22,12 +23,28 @@
...
+    def __init__(self):
+        self._data = {}
+
+    def get_task(self, task_id):
+        return self._data.get(task_id)
+
+    def put_task(self, task):
+        self._data[task.id] = task
+
+    def delete_task(self, task_id):  # pragma: no cover
```
Comment (nuschk) [inline psq/queue_test.py]: Done.

File: psq/task.py (hunk 1 above)
Comment (nuschk) [inline psq/task.py]: Done as wel..

File: psq/queue.py (hunk 2 above)
Comment (nuschk) [inline psq/queue.py]: Oh, and I don't think it'll collide with python3.5 (doesn't in the unit tests, at least). I think python's keyword is in fact `async def` (always in compound). 

It's a bit scary that `async` is colored differently in some editors, though. Then again, RQ does this the same way, so I think we should do that, too.

Comment (jonparrott) [inline psq/queue.py]: Sounds good, just wanted to make sure.

File: psq/queue.py [hunk 5]
```diff
@@ -96,12 +96,15 @@ def enqueue_task(self, task):
         """
         data = dumps(task)
 
-        if self.async:
+        if self._async:
             self.topic.publish(data)
+            logger.info("Task {} queued.".format(task.id))
```
Comment (jonparrott) [inline psq/queue.py]: please use single quotes throughout to maintain consistency in this library.

**Comment by jonparrott** [comment]
@nuschk this is looking good, can you address my final nit (Use single quotes) and get travis to pass?

**Comment by nuschk** [comment]
Aha, the linter fails on an import line being to long, I thought that was ok with the google style guide. Well, no problem, I'll use parentheses, I hope that's ok.

About the double quotes: There are various places where you've already used double quotes for logging statements (e.g., queue.py:76), should I go ahead and change them as well?

**Comment by jonparrott** [comment]
> About the double quotes: There are various places where you've already used double quotes for logging statements (e.g., queue.py:76), should I go ahead and change them as well?

Nah.

**Comment by jonparrott** [comment]
Thanks for your contribution, @nuschk!

**Comment by nuschk** [comment]
Cool, thank you! 👍

**Pull request merged**